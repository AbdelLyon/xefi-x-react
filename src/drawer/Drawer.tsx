import type { JSX, ReactNode } from "react";
import {
  Drawer as DrawerRoot,
  DrawerContent,
  DrawerHeader,
  DrawerBody,
  DrawerFooter,
  type DrawerProps as DrawerRootProps,
} from "@heroui/react";
import { mergeTailwindClasses } from "@/utils";
import { Button, type ButtonProps } from "@/button";
import { useDisclosure } from "@/hooks";

interface DrawerClassNames {
  wrapper?: string;
  base?: string;
  backdrop?: string;
  closeButton?: string;
  header?: string;
  body?: string;
  footer?: string;
}

interface AdditionalDrawerProps {
  trigger?: ReactNode;
  title?: ReactNode;
  children: ReactNode;
  footer?: ReactNode;
  buttonCloseLabel?: string;
  buttonActionLabel?: string;
  onAction?: () => void | Promise<void>;
  buttonCloseProps?: ButtonProps;
  buttonActionProps?: ButtonProps;
  classNames?: DrawerClassNames;
}

export type DrawerProps = Omit<DrawerRootProps, keyof AdditionalDrawerProps> &
  AdditionalDrawerProps;

const isValidButtonLabel = (label: unknown): label is string =>
  typeof label === "string" && label.length > 0;

export const Drawer = ({
  trigger,
  title,
  children,
  footer,
  buttonCloseLabel = "Close",
  buttonActionLabel,
  onAction,
  buttonCloseProps,
  buttonActionProps,
  classNames = {},
  ...props
}: DrawerProps): JSX.Element => {
  const { onOpen, onClose, isOpen } = useDisclosure();

  const handleAction = async (): Promise<void> => {
    try {
      await onAction?.();
      onClose();
    } catch (error) {
      console.error("Action failed:", error);
    }
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>): void => {
    if (event.key === "Enter" || event.key === " ") {
      event.preventDefault();
      onOpen();
    }
  };

  const renderButtons = (): ReactNode => {
    const hasValidCloseLabel = isValidButtonLabel(buttonCloseLabel);
    const hasValidActionButton =
      isValidButtonLabel(buttonActionLabel) && onAction !== undefined;

    const defaultButtonProps = {
      color: "primary" as const,
      radius: "sm" as const,
    };

    return (
      <div className="flex justify-end gap-2">
        {hasValidCloseLabel && (
          <Button
            {...defaultButtonProps}
            variant="bordered"
            onClick={onClose}
            className={mergeTailwindClasses(
              "border-primary/50",
              buttonCloseProps?.className,
            )}
            {...buttonCloseProps}
          >
            {buttonCloseLabel}
          </Button>
        )}

        {hasValidActionButton && (
          <Button
            {...defaultButtonProps}
            onClick={handleAction}
            {...buttonActionProps}
          >
            {buttonActionLabel}
          </Button>
        )}
      </div>
    );
  };

  const drawerClassNames = {
    wrapper: mergeTailwindClasses(classNames.wrapper),
    base: mergeTailwindClasses("bg-background rounded-none", classNames.base),
    backdrop: mergeTailwindClasses(classNames.backdrop),
    closeButton: mergeTailwindClasses(
      "absolute right-4 top-4",
      classNames.closeButton,
    ),
    header: mergeTailwindClasses(classNames.header),
    body: mergeTailwindClasses(classNames.body),
    footer: mergeTailwindClasses(classNames.footer),
  };

  return (
    <>
      {trigger ? (
        <div
          role="button"
          tabIndex={0}
          onClick={onOpen}
          onKeyDown={handleKeyDown}
          className="inline-block"
        >
          {trigger}
        </div>
      ) : null}

      <DrawerRoot
        isOpen={isOpen}
        onClose={onClose}
        classNames={drawerClassNames}
        {...props}
      >
        <DrawerContent>
          {(): JSX.Element => (
            <>
              {title !== undefined && (
                <DrawerHeader className={drawerClassNames.header}>
                  {title}
                </DrawerHeader>
              )}

              <DrawerBody className={drawerClassNames.body}>
                {children}
              </DrawerBody>

              <DrawerFooter className={drawerClassNames.footer}>
                {footer !== undefined ? footer : renderButtons()}
              </DrawerFooter>
            </>
          )}
        </DrawerContent>
      </DrawerRoot>
    </>
  );
};
