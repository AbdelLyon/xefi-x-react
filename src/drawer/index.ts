import "@/index.css";
export { Drawer } from "@/drawer/Drawer";
export type { DrawerProps } from "@/drawer/Drawer";
