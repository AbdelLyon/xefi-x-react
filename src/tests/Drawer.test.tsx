import { Drawer } from "@/drawer";
import { render, screen, fireEvent, act } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";

describe("Composant Drawer", (): void => {
  const defaultProps = {
    trigger: <button>Ouvrir</button>,
    children: <div>Contenu du drawer</div>,
  };

  describe("Snapshots", (): void => {
    it("devrait correspondre au snapshot avec les props par défaut", (): void => {
      const { container } = render(<Drawer {...defaultProps} />);
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot avec un titre", (): void => {
      const { container } = render(
        <Drawer {...defaultProps} title="Titre du drawer" />,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot avec un footer personnalisé", (): void => {
      const { container } = render(
        <Drawer {...defaultProps} footer={<div>Footer personnalisé</div>} />,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot avec des boutons personnalisés", (): void => {
      const { container } = render(
        <Drawer
          {...defaultProps}
          buttonCloseLabel="Fermer"
          buttonActionLabel="Valider"
          onAction={(): void => {}}
        />,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot quand le drawer est ouvert", async (): Promise<void> => {
      const { container } = render(<Drawer {...defaultProps} />);
      await act(async (): Promise<void> => {
        await fireEvent.click(screen.getByText("Ouvrir"));
      });
      expect(container).toMatchSnapshot();
    });
  });

  describe("Comportement", (): void => {
    it("devrait appeler onAction quand le bouton d'action est cliqué", async (): Promise<void> => {
      const onActionMock = vi.fn();
      render(
        <Drawer
          {...defaultProps}
          buttonActionLabel="Valider"
          onAction={onActionMock}
        />,
      );
      await act(async (): Promise<void> => {
        await fireEvent.click(screen.getByText("Ouvrir"));
      });
      await act(async (): Promise<void> => {
        await fireEvent.click(screen.getByText("Valider"));
      });
      expect(onActionMock).toHaveBeenCalledTimes(1);
    });
  });

  describe("Accessibilité", (): void => {
    it("devrait avoir un rôle de bouton pour le déclencheur", (): void => {
      render(<Drawer {...defaultProps} />);
      const trigger = screen.getByText("Ouvrir").closest("div");
      expect(trigger).toHaveAttribute("role", "button");
    });

    it("devrait pouvoir être ouvert avec la touche Entrée", async (): Promise<void> => {
      render(<Drawer {...defaultProps} />);
      const trigger = screen.getByText("Ouvrir").closest("div");
      if (trigger) {
        await act(async (): Promise<void> => {
          await fireEvent.keyDown(trigger, { key: "Enter" });
        });
        expect(screen.getByText("Contenu du drawer")).toBeInTheDocument();
      }
    });
  });
});
