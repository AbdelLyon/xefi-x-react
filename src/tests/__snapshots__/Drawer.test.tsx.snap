// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Composant Drawer > Snapshots > devrait correspondre au snapshot avec des boutons personnalisés 1`] = `
<div>
  <div
    class="inline-block"
    role="button"
    tabindex="0"
  >
    <button>
      Ouvrir
    </button>
  </div>
</div>
`;

exports[`Composant Drawer > Snapshots > devrait correspondre au snapshot avec les props par défaut 1`] = `
<div>
  <div
    class="inline-block"
    role="button"
    tabindex="0"
  >
    <button>
      Ouvrir
    </button>
  </div>
</div>
`;

exports[`Composant Drawer > Snapshots > devrait correspondre au snapshot avec un footer personnalisé 1`] = `
<div>
  <div
    class="inline-block"
    role="button"
    tabindex="0"
  >
    <button>
      Ouvrir
    </button>
  </div>
</div>
`;

exports[`Composant Drawer > Snapshots > devrait correspondre au snapshot avec un titre 1`] = `
<div>
  <div
    class="inline-block"
    role="button"
    tabindex="0"
  >
    <button>
      Ouvrir
    </button>
  </div>
</div>
`;

exports[`Composant Drawer > Snapshots > devrait correspondre au snapshot quand le drawer est ouvert 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="inline-block"
    role="button"
    tabindex="0"
  >
    <button>
      Ouvrir
    </button>
  </div>
</div>
`;
